LibraBFT/Impl/Consensus/BlockStorage/SyncManager.agda

```diff
@@ -4,20 +4,63 @@
    Licensed under the Universal Permissive License v 1.0 as shown at https://opensource.oracle.com/licenses/upl
 -}
 
-open import LibraBFT.Base.ByteString
-open import LibraBFT.Base.PKCS
-open import LibraBFT.Base.Types
-open import LibraBFT.Hash
-open import LibraBFT.Impl.Consensus.ConsensusTypes.Vote as Vote
-open import LibraBFT.ImplShared.Base.Types
+open import LibraBFT.Impl.Consensus.BlockStorage.BlockStore   as BlockStore
+import      LibraBFT.Impl.Consensus.BlockStorage.BlockTree    as BlockTree
+open import LibraBFT.Impl.Consensus.ConsensusTypes.Vote       as Vote
+open import LibraBFT.Impl.OBM.Logging.Logging
 open import LibraBFT.ImplShared.Consensus.Types
-open import LibraBFT.ImplShared.Util.Crypto
 open import LibraBFT.ImplShared.Util.Util
-open import LibraBFT.Abstract.Types.EpochConfig UID NodeId
 open import LibraBFT.Prelude
 open import Optics.All
 
 module LibraBFT.Impl.Consensus.BlockStorage.SyncManager where
 
+data NeedFetchResult : Set where
+  QCRoundBeforeRoot : NeedFetchResult
+  QCAlreadyExist    : NeedFetchResult
+  QCBlockExist      : NeedFetchResult
+  NeedFetch         : NeedFetchResult
+
 postulate
-  insertQuorumCertM : QuorumCert → BlockRetriever → LBFT (Either ErrLog Unit)
+
+  liNextEpochState : Lens LedgerInfo (Maybe EpochState)
+  qcEndsEpoch      : Lens QuorumCert Bool
+
+  fetchQuorumCertM
+    : QuorumCert → BlockRetriever
+    → LBFT (Either ErrLog Unit)
+
+  needFetchForQuorumCert
+    : QuorumCert → BlockStore
+    → Either ErrLog NeedFetchResult
+
+------------------------------------------------------------------------------
+
+insertQuorumCertM
+  : QuorumCert → BlockRetriever
+  → LBFT (Either ErrLog Unit)
+insertQuorumCertM qc retriever = do
+  bs ← use lBlockStore
+  _ ← case needFetchForQuorumCert qc bs of \where
+    (Left e) →
+      bail e
+    (Right NeedFetch) →
+      fetchQuorumCertM qc retriever
+      ∙^∙ withErrCtxt
+    (Right QCBlockExist) →
+      BlockStore.insertSingleQuorumCertM qc ∙^∙ withErrCtxt ∙?∙ λ _ → do
+      ok unit
+    (Right _) →
+      ok unit
+  maybeS (bs ^∙ bsRoot) (bail fakeErr) $ λ bsr →
+    if-dec (bsr ^∙ ebRound) <?ℕ (qc ^∙ qcCommitInfo ∙ biRound)
```

Comment:
DONE.

Reply:
@cwjnkins There are several existing places that use `if-dec` in `LBFT`.
Should they all be changed to `ifM`?

Reply:
@haroldcarr In general, only use `if` when the boolean expression we are inspecting is _exactly_ a variable that has been bound by a monad bind; otherwise use `ifM`.

Reply:
Understood - if the are in a monad bind, then they are "in" the djistra monad --- otherwise `ifM` will put them there.
I updated two places.